nova/compute/api.py

```diff
@@ -1773,10 +1774,20 @@ def soft_delete(self, context, instance):
         """Terminate an instance."""
         LOG.debug('Going to try to soft delete instance',
                   instance=instance)
-
-        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
-                     task_state=task_states.SOFT_DELETING,
-                     deleted_at=timeutils.utcnow())
+        if instance['vm_state'] == vm_states.SOFT_DELETED:
+            LOG.warn("%s is soft delete state" % (instance.uuid))
+            reason = ("The instance is %s state" % (instance.vm_state))
+            raise exception.InstanceTerminationFailure(reason=reason)
+        elif instance['vm_state'] == vm_states.ACTIVE or \
+                instance['vm_state'] == vm_states.STOPPED:
+            self._delete(context, instance, 'soft_delete',
```

Comment:
代码中已经全部放开软删除操作．代码已提交．

2016-06-24 16:25 GMT+08:00 blkart <notifications@github.com>:

> In nova/compute/api.py
> <https://github.com/eayunstack/nova/pull/37#discussion_r68367808>:
>
> > @@ -1773,10 +1774,20 @@ def soft_delete(self, context, instance):
> >          """Terminate an instance."""
> >          LOG.debug('Going to try to soft delete instance',
> >                    instance=instance)
> > -
> > -        self._delete(context, instance, 'soft_delete', self._do_soft_delete,
> > -                     task_state=task_states.SOFT_DELETING,
> > -                     deleted_at=timeutils.utcnow())
> > +        if instance['vm_state'] == vm_states.SOFT_DELETED:
> > +            LOG.warn("%s is soft delete state" % (instance.uuid))
> > +            reason = ("The instance is %s state" % (instance.vm_state))
> > +            raise exception.InstanceTerminationFailure(reason=reason)
> > +        elif instance['vm_state'] == vm_states.ACTIVE or \
> > +                instance['vm_state'] == vm_states.STOPPED:
> > +            self._delete(context, instance, 'soft_delete',
>
> 只允许删除 ACTIVE 和 STOPPED 状态的虚拟机是否妥当
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eayunstack/nova/pull/37/files/77d779d01ef86664194b0ef8116f62792a28ceb3#r68367808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKMiZkqEkVaItCL83xD3eBxLM3EHMpkAks5qO5R-gaJpZM4I0Frh>
> .
>